keep-network/keep-ecdsa
# Proof of Concept: Liquidation Recovery

This PR makes calls to the newly merged liquidation recovery functions in order to construct a signed bitcoin transaction to recover the funds.

This PR does not implement persistence for address validation - that'll be separate!

To test:

https://github.com/keep-network/local-setup/blob/master/docs/manual-setup.adoc

Edit your `keep-ecdsa/configs/config.local.*.toml` files to have:
```
[Extensions.TBTC]
	TBTCSystem = "<value matching ethereum.ContractAddresses.TBTCSystem>"

  [Extensions.TBTC.BTCRefunds]
  	BeneficiaryAddress = "<either a valid btc address or a HD wallet key>"
  	MaxFeePerVByte = <integer>
```

* Get keep-core, and keep-ecdsa 1-3 running locally.
* from the `tbtc.js` project run:
```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit new 10000000
```

That should give you a bitcoin address to make a deposit to:
```
Got deposit address: bcrt1q7h3eg2vn9npa4szyanwk9m82vgz7n4jdnke0cj ; fund with: 10000000 satoshis please.
```

Make the deposit with:
```
$ bitcoind-wallet sendToAddress bcrt1q7h3eg2vn9npa4szyanwk9m82vgz7n4jdnke0cj 0.1
```

Once that clears, from the `tbtc/solidity/scripts` directory run:
```
$ truffle exec multiply-local-eth-btc-price.js 0.7
```
That'll lower the price ratio of eth and btc enough to under-collateralize the deposit and make it available for liquidation.

back in `tbtc.js`, run:
```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit list --vending-machine
```
and look for a line like:
```
<hex-to-liquidate>	ACTIVE	10000000	104
```

Grab your hex address and then run
```
$ bin/tbtc.js --rpc ws://127.0.0.1:8546 deposit <hex-to-liquidate> liquidate
```

Now, you can monitor your ecdsa logs, looking out for something that looks like:
![image](https://user-images.githubusercontent.com/1045160/112218320-f5aff880-8bf9-11eb-8204-27f16935852c.png)

repeated 5x

This is still a work in progress, but any early review would be valuable, @nkuba 

:tada:
